### ISSUE TITLE
Unable to create agents from agents

### ISSUE THREAD

**Issue by one4many** [opensistemas-hub/osbrain#100]
I'm trying to write an agentherder which allows one agent (in charge) to create new ones under his her command. This fails with a Pyro "AssertionError: daemonic processes are not allowed to have children" which seems a bit extreme not allowing sub processes.

Example code in here
[ex.txt](https://github.com/opensistemas-hub/osbrain/files/876005/ex.txt)

Can this be fixed? Or done another way with osbrain?

**Comment by Peque** [comment]
Thanks for reporting this issue.

We were aware of this problem. Actually, we tried to make the name server do exactly that (it seemed like a reasonable feature). As we did not have an actual use case at the moment, we did not investigate further.

We will investigate it further and try to fix it for the (hopefully close) 1.0.0 version.

**Comment by one4many** [comment]
Very much appreciated. I'm currently working on a project which depends on this feature. A workaround would cause us a major headache and lots of (unnecessary) lines of code.

**Comment by Flood1993** [comment]
I have been playing a bit with this, and it seems to work if you execute the program with the `-O` flag.

For example, try [example.txt](https://github.com/opensistemas-hub/osbrain/files/884865/example.txt) by executing it with `$ python -O ...`

However, the shutdown does not seem to be working correctly, my guess being that because of the child process not being a direct children of the "main" python process.

Snapshot of process when executing the code above:
```
$ ps -l -a
F S   UID   PID  PPID  C PRI  NI ADDR SZ WCHAN  TTY          TIME CMD
0 S  1001  2324  2323  0  80   0 -  6940 wait   pts/5    00:00:00 bash
0 S  1001  4933  2324  4  80   0 - 34141 wait   pts/5    00:00:00 python
1 S  1001  4937  4933  0  80   0 - 162704 inet_c pts/5   00:00:00 python
1 S  1001  4945  4933  0  80   0 - 183654 inet_c pts/5   00:00:00 python
1 S  1001  4955  4945  0  80   0 - 185703 inet_c pts/5   00:00:00 python
0 R  1001  4966  3925  0  80   0 -  3577 -      pts/6    00:00:00 ps
```

**Comment by one4many** [comment]
Interesting ... for me the shutdown via parent process does not work, only direct. See
[ex3.txt](https://github.com/opensistemas-hub/osbrain/files/885288/ex3.txt)


```
alex@dev:~/osb$ python -O  herderExample.py 
Not starting broadcast server for localhost.
NS running on 127.0.0.1:19921 (127.0.0.1)
URI = PYRO:Pyro.NameServer@127.0.0.1:19921
Working for the president
Working for her majesty
Working for her majesty
b
b1
INFO [2017-03-31 10:26:12.336584] (Agent007): Stopping...
Working for our king
a
b
Agent001 <- Working for the president
Agent007 <- Working for her majesty
Agent007 <- nothing to report
Agent007 <- got shutdown order
Agent004 <- Working for our king
```
The output stops after `b` it seems the `send` hangs. Removing it does not get me any further because the `shutdown()` hangs next.

**Comment by Flood1993** [comment]
Thanks for your feedback. Would you mind to reduce the code as much as you can to a minimum, reproducible case? Also, take a look at #107 . I will answer again after I've made some more research.

**Comment by one4many** [comment]
A stripped down version of [ex3.txt](https://github.com/opensistemas-hub/osbrain/files/889608/ex3.txt)

Looked at #107 but hit some issues (see comments there). I personally prefer to write my own sub/childagent class.

**Comment by one4many** [comment]
Any updates? I have a hard stop on my project here because of this (missing) feature.

**Comment by Peque** [comment]
@one4many Thanks for your patience. We will be updating this issue with any changes we make, so you will be notified once we implement it.

We do not forget about it, you can see this issue is in our [in-progress task list](https://github.com/opensistemas-hub/osbrain/projects/1).

**Comment by Peque** [comment]
@one4many I have found the problem. A fix should be available very soon in a development branch, although it will not be merged into master and officially released until next week (@Flood1993 is on holidays and I will wait for his review before merging and releasing). :wink:

**Comment by Peque** [comment]
Pull request proposed (#112).

@one4many You may try those changes using the `inception` branch available in my fork (https://github.com/Peque/osbrain/tree/inception).

**Comment by Peque** [comment]
Merged and included in [0.4.3 release](https://pypi.python.org/pypi?:action=display&name=osbrain&version=0.4.3).

**Comment by one4many** [comment]
Works for me ... Thank you @Peque @Flood1993
